File: 2018_03_01/NorthwindRest/NorthwindRest.NetWebApi.VB/Controllers/CategoriesController.vb
Imports System.Net
Imports System.Web.Http
Imports Newtonsoft.Json
Imports NorthwindRest.DataAccess

<RoutePrefix("api/categories")>
Public Class CategoriesController
    Inherits ApiController

    ' GET api/<controller>
    <HttpGet, Route("all")>
    Public Function GetValues() As IEnumerable(Of Category)
        Dim db As New NorthwindDb()
        db.Configuration.LazyLoadingEnabled = False
        db.Configuration.ProxyCreationEnabled = False
        Dim repo As New GenericRepository(Of Category)(db)
        Dim result = repo.Find().ToList()

        ' Verhindert zirkuläre Bezüge bei Serialisierung
        Dim settings As JsonSerializerSettings = New JsonSerializerSettings() With
        {.ObjectCreationHandling = ObjectCreationHandling.Auto, .PreserveReferencesHandling = PreserveReferencesHandling.Objects}
        Dim json = JsonConvert.SerializeObject(result, settings)
        Return result
    End Function

    ' GET api/<controller>/5
    Public Function GetValue(ByVal id As Integer) As String
        Return "value"
    End Function

    ' POST api/<controller>
    Public Sub PostValue(<FromBody()> ByVal value As String)

    End Sub

    ' PUT api/<controller>/5
    Public Sub PutValue(ByVal id As Integer, <FromBody()> ByVal value As String)

    End Sub

    ' DELETE api/<controller>/5
    Public Sub DeleteValue(ByVal id As Integer)

    End Sub
End Class

File: 2018_03_01/NorthwindRest/NorthwindRest.NetWebApi/Controllers/CategoriesController.cs
using NorthwindRest.DataAccess;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;
using System.Web.Http;
using System.Web.Http.Routing;

namespace NorthwindRest.NetWebApi.Controllers
{
    [RoutePrefix("api/categories")]
    public class CategoriesController : ApiController
    {
        [HttpGet, Route("all")]
        public async Task<IHttpActionResult> GetAll()
        {
            var db = new NorthwindDb();
            // Dann sind die Produkte nicht mehr gefüllt
            // db.Configuration.LazyLoadingEnabled = false;
            // Dann werden echte Category-Objekte (kleiner) verwendet
            db.Configuration.ProxyCreationEnabled = false;
            var repo = new GenericRepository<Category>(db);
            var result = await Task.Run<IEnumerable<Category>>(()=> repo.Find().ToList());
            db.Dispose();
            return Ok(result);
        }

        [HttpGet, Route("{id}")]
        public string Get(int id)
        {
            return "value";
        }

        [HttpPost, Route("new")]
        public IHttpActionResult Post([FromBody]Category value)
        {
            value.CategoryId = 9;

            var link = Url.Link("DefaultApi", 
                new { controller = "Categories", id = value.CategoryId });
            return Created(link, value);
        }
    }
}